Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialAsset.cpp

```diff
@@ -109,6 +109,81 @@ namespace AZ
             return m_wasPreFinalized;
         }
 
+        //! Attempts to convert a numeric MaterialPropertyValue to another numeric type @T,
+        //! since MaterialPropertyValue itself does not support any kind of casting.
+        //! If the original MaterialPropertyValue is not a numeric type, the original value is returned.
+        template<typename T>
+        MaterialPropertyValue CastNumericMaterialPropertyValue(const MaterialPropertyValue& value)
+        {
+            TypeId typeId = value.GetTypeId();
+
+            if (typeId == azrtti_typeid<bool>())
+            {
+                return aznumeric_cast<T>(value.GetValue<bool>());
+            }
+            else if (typeId == azrtti_typeid<int32_t>())
+            {
+                return aznumeric_cast<T>(value.GetValue<int32_t>());
+            }
+            else if (typeId == azrtti_typeid<uint32_t>())
+            {
+                return aznumeric_cast<T>(value.GetValue<uint32_t>());
+            }
+            else if (typeId == azrtti_typeid<float>())
+            {
+                return aznumeric_cast<T>(value.GetValue<float>());
+            }
+            else
+            {
+                return value;
+            }
+        }
+        
+        //! Attempts to convert an AZ::Vector[2-4] MaterialPropertyValue to another AZ::Vector[2-4] type @T.
+        //! Any extra elements will be dropped or set to 0.0 as needed.
+        //! If the original MaterialPropertyValue is not a Vector type, the original value is returned.
+        template<typename VectorT>
+        MaterialPropertyValue CastVectorMaterialPropertyValue(const MaterialPropertyValue& value)
```

Comment:
Should this support converting a single floating point value to a vector or vice versa?

Reply:
My goal was just to replace whatever automatic conversions work in the JSON serialization system, and that doesn't do conversion between float and Vector types.